browser/ui/views/location_bar/brave_search_conversion/promotion_button_controller.cc

```diff
@@ -0,0 +1,237 @@
+  button_ = nullptr;
+  omnibox_view_ = nullptr;
+}
+
+void PromotionButtonController::Show(bool show) {
+  button_->SetVisible(show);
+
+  // For now, animation is triggered only once after changing search provider.
+  // Showing animation whenever this button is visible is too much.
+  // TODO(simonhong): adjust how frequently this animation is used.
+  if (use_animation_ && show) {
+    button_->AnimateExpand();
+    use_animation_ = false;
+  }
+}
+
+bool PromotionButtonController::ShouldShowSearchPromotionButton() {
+  if (!button_ || !omnibox_view_) {
+    return false;
+  }
+
+  if (is_brave_search_default_) {
+    return false;
+  }
+
+  if (prefs_->GetBoolean(brave_search_conversion::prefs::kDismissed)) {
+    return false;
+  }
+
+  // No popup means no suggestions for current input.
+  // Promotion button will be shown for current search provider's
+  // suggestion entries to make users search with brave search with that
+  // suggestion.
+  if (!omnibox_view_->model()->PopupIsOpen()) {
+    return false;
+  }
+
+  // Only show promotion for search query. Not url.
+  if (omnibox_view_->model()->CurrentTextIsURL()) {
+    return false;
+  }
+
+  const AutocompleteMatch match = omnibox_view_->model()->CurrentMatch(nullptr);
+  return !IsBraveSearchPromotionMatch(match) &&
+#if BUILDFLAG(ENABLE_AI_CHAT)
+         !LeoProvider::IsMatchFromLeoProvider(match) &&
+#endif
+         AutocompleteMatch::IsSearchType(match.type);
+}
+
+void PromotionButtonController::OnViewIsDeleting(views::View* observed_view) {
+  // If any observed view among butto or omnibox is destroying,
+  // this controller will not do anything after that.
+  view_observation_.RemoveAllObservations();
+  button_ = nullptr;
+  omnibox_view_ = nullptr;
+}
+
+void PromotionButtonController::OnTemplateURLServiceChanged() {
+  use_animation_ = true;
+  is_brave_search_default_ = IsBraveSearchDefault();
+  if (is_brave_search_default_) {
+    return;
+  }
+
+  UpdateButtonUI();
+}
+
+void PromotionButtonController::OnTemplateURLServiceShuttingDown() {
+  template_url_service_observation_.Reset();
+}
+
+void PromotionButtonController::SetDefaultAndLoadBraveSearchWithCurrentInput() {
+  CHECK(omnibox_view_);
+
+  // Set brave search as default.
+  auto provider_data = TemplateURLDataFromPrepopulatedEngine(
+      TemplateURLPrepopulateData::brave_search);
+  TemplateURL template_url(*provider_data);
+  template_url_service_->SetUserSelectedDefaultSearchProvider(&template_url);
+
+  // Load brave search with current input.
+  const auto url =
+      template_url_service_->GenerateSearchURLForDefaultSearchProvider(
+          omnibox_view_->GetText());
+
+  NavigateParams params(&*browser_, url, ui::PAGE_TRANSITION_TYPED);
+  params.disposition = WindowOpenDisposition::CURRENT_TAB;
+  Navigate(&params);
+}
+
+void PromotionButtonController::Dismissed() {
+  button_->SetVisible(false);
+  brave_search_conversion::SetDismissed(&*prefs_);
```

Comment:
Should this have a more specific name? I think we've got a few different conversion methods, which presumably use different prefs?

`SetDismissedLocationBarButton` or something?

Reply:
This is intended as user dismissed already via another promotion, I think we don't want to show this button again.
@rebron WDYT?

Reply:
Okay, that makes sense :)